File: client/result_tree_test.go
package client

import (
	"testing"

	"github.com/davecgh/go-spew/spew"
	"github.com/graphql-go/graphql/language/ast"
	"github.com/graphql-go/graphql/language/parser"
	"github.com/rgraphql/magellan/schema"
	"github.com/rgraphql/magellan/types"
	proto "github.com/rgraphql/rgraphql"
)

func buildTestSchema(t *testing.T) *schema.Schema {
	scm, err := schema.Parse(`
	type Person {
		name: String
		height: Int
	}

	type RootQuery {
		allPeople: [Person]
		names: [String]
		person(name: String): Person!
	}

	schema {
		query: RootQuery
	}
	`)
	if err != nil {
		t.Fatal(err.Error())
	}
	return scm
}

// TestSimpleApply is a basic preliminary test of the result tree.
func TestSimpleApply(t *testing.T) {
	scm := buildTestSchema(t)

	pdoc, err := parser.Parse(parser.ParseParams{
		Source: `
		query {
			person(name: "Christian") {
				name
			}
		}
		query {
			person(name: "Charlie") {
				name
			}
		}
		`,
	})
	if err != nil {
		t.Fatal(err.Error())
	}

	qtree, err := NewQueryTree(scm, nil)
	if err != nil {
		t.Fatal(err.Error())
	}

	query, err := NewQuery(pdoc.Definitions[0].(*ast.OperationDefinition))
	if err != nil {
		t.Fatal(err.Error())
	}

	err = qtree.Attach(query)
	if err != nil {
		t.Fatal(err.Error())
	}

	rtree, err := NewResultTree(qtree, proto.RGQLValueInit_CACHE_LRU, 30)
	if err != nil {
		t.Fatal(err.Error())
	}

	vals := []*proto.RGQLValue{
		{QueryNodeId: 1},
		{ArrayIndex: 1},
		{PosIdentifier: 1, QueryNodeId: 2, Value: types.NewStringPrimitive("hello world")},
		{PosIdentifier: 1, Value: types.NewStringPrimitive("goodbye")},
	}
	for vali, val := range vals {
		if err := rtree.HandleValue(val); err != nil {
			t.Fatalf("val[%d]: %s", vali, err.Error())
		}
	}

	t.Logf("applied %d mutations", len(vals))
	spew.Dump(rtree.root)
	// t.Logf("output: %#v", rtree.root)
}

File: client/result-tree.test.ts
import { parse, buildSchema, OperationDefinitionNode } from 'graphql'
import { QueryTree } from './query-tree.js'
import { QueryTreeHandler } from './query-tree-handler.js'
import { ResultTree } from './result-tree.js'
import { PackPrimitive } from '../primitive.js'
import {
  DeepPartial,
  RGQLQueryTreeMutation,
  RGQLValue,
  RGQLValueInit_CacheStrategy,
} from '../rgraphql.pb.js'

function mockSchema() {
  return buildSchema(`
  type RootQuery {
      person: Person
  }

  type Person {
    name: String
    age: Int
    description: String
  }

  schema {
    query: RootQuery
  }
  `)
}

function mockAst() {
  return parse(
    `query myQuery {
  person {
    name
  }
}
query mySecondQuery {
  person {
    description
  }
}
`,
  )
}

describe('QueryTreeNode', () => {
  it('should purge result tree nodes correctly', () => {
    const queryAst = mockAst()
    const handler: QueryTreeHandler = (mutation: RGQLQueryTreeMutation) => {
      console.log('Applying:')
      console.log(mutation)
    }
    const schema = mockSchema()
    const tree = new QueryTree(schema, handler)
    const rtree = new ResultTree(tree, RGQLValueInit_CacheStrategy.CACHE_LRU, 100)

    const querya = tree.buildQuery(queryAst.definitions[0] as OperationDefinitionNode, {})
    const queryb = tree.buildQuery(queryAst.definitions[1] as OperationDefinitionNode, {})
    // expect(tree.children.length).toBe(3)

    const vals: DeepPartial<RGQLValue>[] = [
      { queryNodeId: 1 },
      { queryNodeId: 2, value: PackPrimitive('test') },
      { queryNodeId: 1 },
      { queryNodeId: 3, value: PackPrimitive('descrip') },
    ]
    for (const val of vals) {
      rtree.handleValue(RGQLValue.create(val))
    }

    const rtRoot = rtree.getRoot()
    expect(rtRoot.children.length).toBe(1)
    expect(rtRoot.children[0].children.length).toBe(2)

    tree.detach(querya)
    expect(rtRoot.children[0].children.length).toBe(1)
    // expect(tree.children.length).toBe(0);
    console.log(queryb)
  })
})
